File: validation/validation-annotation-processor/src/main/java/ru/tinkoff/kora/validation/annotation/processor/extension/ValidKoraExtension.java
package ru.tinkoff.kora.validation.annotation.processor.extension;

import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import jakarta.annotation.Nullable;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.Set;

import static ru.tinkoff.kora.validation.annotation.processor.ValidTypes.VALIDATOR_TYPE;
import static ru.tinkoff.kora.validation.annotation.processor.ValidTypes.VALID_TYPE;

public final class ValidKoraExtension implements KoraExtension {

    private final Types types;
    private final Elements elements;

    public ValidKoraExtension(ProcessingEnvironment processingEnv) {
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, Set<String> tags) {
        if (!tags.isEmpty()) return null;
        if (!(TypeName.get(typeMirror) instanceof ParameterizedTypeName ptn)) {
            return null;
        }
        if (!ptn.rawType().equals(VALIDATOR_TYPE)) {
            return null;
        }
        if (!(typeMirror instanceof DeclaredType dt)) {
            return null;
        }
        var validatorArgumentType = dt.getTypeArguments().get(0);
        if (validatorArgumentType.getKind() != TypeKind.DECLARED) {
            return null;
        }

        var validatedTypeElement = types.asElement(validatorArgumentType);
        if (AnnotationUtils.findAnnotation(validatedTypeElement, VALID_TYPE) == null) {
            return null;
        }
        return KoraExtensionDependencyGenerator.generatedFrom(elements, validatedTypeElement, VALIDATOR_TYPE);
    }
}

File: validation/validation-symbol-processor/src/main/kotlin/ru/tinkoff/kora/validation/symbol/processor/extension/ValidKoraExtension.kt
package ru.tinkoff.kora.validation.symbol.processor.extension

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.validation.symbol.processor.ValidTypes.VALIDATOR_TYPE
import ru.tinkoff.kora.validation.symbol.processor.ValidTypes.VALID_TYPE

class ValidKoraExtension() : KoraExtension {

    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: String?): (() -> ExtensionResult)? {
        if (tag != null) return null
        type.toTypeName().let {
            if (it is ParameterizedTypeName && it.rawType == VALIDATOR_TYPE) {
                val validTypeDecl = type.arguments.first().type?.resolve()?.declaration
                if (validTypeDecl?.isAnnotationPresent(VALID_TYPE) == true) {
                    return generatedByProcessor(resolver, validTypeDecl as KSClassDeclaration, VALIDATOR_TYPE)
                }
            }
        }
        return null
    }
}
